src/ophyd_async/core/_device.py

```diff
@@ -89,29 +90,38 @@ async def connect(
             Time to wait before failing with a TimeoutError.
         """
 
+        if mock is True:  # create a new Mock if one not provided
+            mock = Mock()
+
         if (
             self._previous_connect_was_mock is not None
-            and self._previous_connect_was_mock != mock
+            and self._previous_connect_was_mock != isinstance(mock, Mock)
         ):
             raise RuntimeError(
-                f"`connect(mock={mock})` called on a `Device` where the previous "
-                f"connect was `mock={self._previous_connect_was_mock}`. Changing mock "
-                "value between connects is not permitted."
+                "connect(mock) was called with mock evaluating "
+                f"to {bool(mock)} when previous connect had mock "
+                f" evaluate to {self._previous_connect_was_mock}. "
+                "Changing mock value between connects is not permitted."
             )
-        self._previous_connect_was_mock = mock
+
+        self._previous_connect_was_mock = isinstance(mock, Mock)
 
         # If previous connect with same args has started and not errored, can use it
         can_use_previous_connect = self._connect_task and not (
             self._connect_task.done() and self._connect_task.exception()
         )
         if force_reconnect or not can_use_previous_connect:
             # Kick off a connection
-            coros = {
-                name: child_device.connect(
-                    mock, timeout=timeout, force_reconnect=force_reconnect
+            coros = {}
+
+            for child_name, child_device in self.children():
+                # if mock is Mock instance, get child Mock if exists
+                # else pass down boolean
+                child_mock = getattr(mock, child_name, True) if mock else mock
```

Comment:
mock should always be a `Mock` here by now
```suggestion
                child_mock = getattr(mock, child_name) if mock else mock
```